Include/Rover/Factor.hpp

```diff
@@ -0,0 +1,186 @@
+
+      //! Registers a new category, if not already added.
+      /*!
+        \param category The category to add.
+        \details Amortized O(1) complexity.
+      */
+      void add_category(const Type& category);
+
+      //! Returns a unique dimension from 0 to (number of categories - 1) for
+      //! a given category. Returns nullopt if the category is not registered.
+      /*!
+        \param category The category.
+        \details Amortized O(1) complexity.
+      */
+      std::optional<std::size_t> get_dimension(const Type& category) const;
+
+    private:
+      std::unordered_map<Type, std::size_t> m_map;
+  };
+
+  //! Maps categories of a categorical variable to different dimensions.
+  /*!
+    \tparam T The type of the categorical variable with no std::hash
+              specialized but operator < defined.
+  */
+  template<typename T>
+  class Factor<T, std::enable_if_t<!Details::is_hashable_v<T> &&
+      Details::is_comparable_v<T>>> {
+    public:
+
+      //! The type of the categorical variable.
+      using Type = T;
+
+      //! Registers a new category, if not already added.
+      /*!
+        \param category The category to add.
+        \details O(log N) complexity.
+      */
+      void add_category(const Type& category);
+      
+      //! Returns a unique dimension from 0 to (number of categories - 1) for
+      //! a given category. Returns nullopt if the category is not registered.
+      /*!
+        \param category The category.
+        \details O(log N) complexity. 
+      */
+      std::optional<std::size_t> get_dimension(const Type& category) const;
+
+    private:
+      std::map<Type, std::size_t> m_map;
+  };
+
+  template<typename T, typename C>
+  void Factor<T, C>::add_category(const Type& category) {
+    if(std::find(m_categories.begin(), m_categories.end(), category) ==
+        m_categories.end()) {
+      m_categories.push_back(category);
+    }
+  }
+
+  template<typename T, typename C>
+  std::optional<std::size_t> Factor<T, C>::get_dimension(const Type& category)
+      const {
+    auto it = std::find(m_categories.begin(), m_categories.end(), category);
+    if(it == m_categories.end()) {
+      return std::nullopt;
+    } else {
+      return it - m_categories.begin();
+    }
+  }
+
+  template<typename T>
+  void Factor<T, std::enable_if_t<Details::is_hashable_v<T>>>::add_category(
+      const Type& category) {
+    m_map.try_emplace(category, m_map.size());
+  }
+
+  template<typename T>
+  std::optional<std::size_t> Factor<T, std::enable_if_t<Details::is_hashable_v<
+      T>>>::get_dimension(const Type& category) const {
+    auto it = m_map.find(category);
+    if(it == m_map.end()) {
+      return std::nullopt;
+    } else {
+      return it->second;
+    }
+  }
+
+  template<typename T>
+  void Factor<T, std::enable_if_t<!Details::is_hashable_v<T> &&
+      Details::is_comparable_v<T>>>::add_category(
+      const Type& category) {
+    auto it = m_map.lower_bound(category);
```

Comment:
Could you instead do:

```
  auto i = m_map.find(category);
  if(i == m_map.end()) {
    m_map.emplace_hint(i, category, m_map.size());
  }
}
```

Reply:
If I use `m_map.end()` as a hint, it is not going to help the insertion since the implementation of `std::map` has two options:
1. Look around the end of the tree, i.e. where the greatest values reside. It is not correct, for example, if the map contains `{ 3, 4, 5, 6 }` and I insert 2.
2. Ignore the hint altogether, which I think most implementations do.
The problem here is that if an element is not found, the information about its expected location is not stored in the iterator since `std::map::end` is always the same thing, no matter what key I used for lookup.
The lower bound, in turn, always provides the implementation with the nearest element, which can be used to make insertions constant-time.

Reply:
Excellent explanation!